contents/javascript/Javascript-framework-Redux.md

```diff
@@ -0,0 +1,226 @@
+
+For example, we have two simpler reducers.
+
+Firstly, a reducer for the visibility filter `visibilityFilter`:
+
+```js
+function visibilityFilter(state = 'SHOW_ALL', action) {
+  if (action.type === 'SET_VISIBILITY_FILTER') {
+    return action.filter
+  } else {
+    return state
+  }
+}
+```
+
+And secondly, a reducer, that manages the todo-list items `todos`:
+
+```js
+function todos(state = [], action) {
+  switch (action.type) {
+    case 'ADD_TODO':
+      return state.concat([{ text: action.text, completed: false }])
+    case 'TOGGLE_TODO':
+      return state.map((todo, index) =>
+        action.index === index
+          ? { text: todo.text, completed: !todo.completed }
+          : todo
+      )
+    default:
+      return state
+  }
+}
+```
+
+These two reducers are called by a main reducer, and with that the complete app state is managed.
+
+```js
+function todoApp(state = {}, action) {
+  return {
+    todos: todos(state.todos, action),
+    visibilityFilter: visibilityFilter(state.visibilityFilter, action)
+  }
+}
+```
+
+The reducers are **pure** functions. They take the previous state and an action, and return the next state, as new state objects, instead of mutating the previous state.
+
+A careful eye would notice that the above code block examples on reducers are not pure function. In fact, with some Redux API, a more proper example would then be the code block below. Don't worry if you don't fully understand the specific Javascript syntax and methods used (such as `.map` and the spread operator), but just understand that new state objects are returns when state is updated.
+
+```js
+function visibilityFilter(state = 'SHOW_ALL', action) {
+  switch (action.type) {
+    case 'SET_VISIBILITY_FILTER':
+      return action.filter
+    default:
+      return state
+  }
+}
+function todos(state = [], action) {
+  switch (action.type) {
+    case 'ADD_TODO':
+      return [
+        ...state,
+        {
+          text: action.text,
+          completed: false
+        }
+      ]
+    case 'COMPLETE_TODO':
+      return state.map((todo, index) => {
+        if (index === action.index) {
+          return Object.assign({}, todo, {
+            completed: true
+          })
+        }
+        return todo
+      })
+    default:
+      return state
+  }
+}
+import { combineReducers, createStore } from 'redux'
+const reducer = combineReducers({ visibilityFilter, todos })
+const store = createStore(reducer)
+```
+
+## Why use Redux?
+
+([Adapted from this artcle](https://blog.logrocket.com/why-use-redux-reasons-with-clear-examples-d21bffd5835/))
+([And this article](https://www.smashingmagazine.com/2016/06/an-introduction-to-redux/))
+
+Now that we know what Redux is, let us take a look at some of its benefits that were made possible with these concepts of Redux. 
+
+1. **State is made predictable**<br/>
```

Comment:
Even as an experienced user, this point was somewhat confusing 😅, especially the second paragraph.
Try to make it simpler and tied to _your_ article, departing from the linked articles as they provide some context before these points.

Reply:
Ok. Help me re-read again.